src/cpp/rtps/builtin/discovery/endpoint/EDP.cpp

```diff
@@ -1410,6 +1472,74 @@ bool EDP::hasTypeObject(const WriterProxyData* wdata, const ReaderProxyData* rda
     return false;
 }
 
+const SubscriptionMatchedStatus EDP::update_subscription_matched_status(
+        GUID_t reader_guid,
+        GUID_t writer_guid,
+        int change)
+{
+    SubscriptionMatchedStatus status;
+    auto it = reader_status_.find(reader_guid);
+    if(it != reader_status_.end())
+    {
+        it->second.current_count += change;
+        it->second.current_count_change += change;
+        it->second.total_count += change;
+        it->second.total_count_change +=change;
+        it->second.last_publication_handle = writer_guid;
+        status = it->second;
+        it->second.current_count = 0;
+        it->second.current_count_change = 0;
+        return status;
+    }
+    else
+    {
+        status.current_count += change;
+        status.current_count_change += change;
+        status.total_count += change;
+        status.total_count_change +=change;
+        status.last_publication_handle = writer_guid;
+        SubscriptionMatchedStatus returned = status;
+        status.current_count = 0;
+        status.current_count_change = 0;
+        reader_status_.emplace(reader_guid, status);
+        return returned;
+    }
+}
```

Comment:
Refactor the whole method with the following guidelines:
* return const ref instead of a copy.
* arguments should be const ref too.
* if the `guid` isn't found, call emplace, this will construct the object directly on the map.
* if the `guid` is found, retrieve the stored value, modify it and return it.

As snipped:

    SubscriptionMatchedStatus& status;
    auto it = reader_status_.find(reader_guid);
    if (it == reader_status_.end())
    {
        auto pair = reader_status_.emplace(reader_guid, SubscriptionMatchedStatus{});
        status = *pair.second;
    }
    else
    {
        status = it->second;
    }

    status.current_count += change;
    status.current_count_change += change;
    status.total_count += change;
    status.total_count_change +=change;
    status.last_publication_handle = writer_guid;

    return status